--- compiler/parsing/syntaxerr.ml	2007-06-05 20:51:48.000000000 +0200
+++ ../compiler/parsing/syntaxerr.ml	2007-06-05 20:51:48.000000000 +0200
@@ -1,4 +1,30 @@
-(* This file was syntaxerr.ml from Objective Caml *)
+(**********************************************************************)
+(*                                                                    *)
+(*                           ReactiveML                               *)
+(*                    http://reactiveML.org                           *)
+(*                    http://rml.inria.fr                             *)
+(*                                                                    *)
+(*                          Louis Mandel                              *)
+(*                                                                    *)
+(*  Copyright 2002, 2007 Louis Mandel.  All rights reserved.          *)
+(*  This file is distributed under the terms of the Q Public License  *)
+(*  version 1.0.                                                      *)
+(*                                                                    *)
+(*  ReactiveML has been done in the following labs:                   *)
+(*  - theme SPI, Laboratoire d'Informatique de Paris 6 (2002-2005)    *)
+(*  - Verimag, CNRS Grenoble (2005-2006)                              *)
+(*  - projet Moscova, INRIA Rocquencourt (2006-2007)                  *)
+(*                                                                    *)
+(**********************************************************************)
+
+(* file: syntaxerr.ml *)
+
+(* Warning: *)
+(* This file is based on the original version of syntaxerr.ml *)
+(* from the Objective Caml 3.07 distribution, INRIA           *)
+
+(* first modification: 2004-05-05 *)
+(* modified by: Louis Mandel      *)
 
 (***********************************************************************)
 (*                                                                     *)
@@ -12,7 +38,7 @@
 (*                                                                     *)
 (***********************************************************************)
 
-(* $Id: syntaxerr.ml,v 1.8 2002/04/18 08:50:43 garrigue Exp $ *)
+(* $Id$ *)
 
 (* Auxiliary type for reporting syntax errors *)
 
@@ -27,16 +53,10 @@
 
 let report_error ppf = function
   | Unclosed(opening_loc, opening, closing_loc, closing) ->
-      if String.length !Location.input_name = 0
-      && Location.highlight_locations ppf opening_loc closing_loc
-      then fprintf ppf "Syntax error: '%s' expected, \
-                   the highlighted '%s' might be unmatched" closing opening
-      else begin
-        fprintf ppf "%aSyntax error: '%s' expected@."
-          Location.print closing_loc closing;
-        fprintf ppf "%aThis '%s' might be unmatched"
-          Location.print opening_loc opening 
-      end
+      fprintf ppf "%aSyntax error: '%s' expected@."
+        Location.print closing_loc closing;
+      fprintf ppf "%aThis '%s' might be unmatched"
+        Location.print opening_loc opening 
   | Other loc ->
       fprintf ppf "%aSyntax error" Location.print loc
 
