--- compiler/global/warnings.ml	2007-12-05 19:01:30.000000000 +0100
+++ ../compiler/global/warnings.ml	2007-12-05 19:01:30.000000000 +0100
@@ -1,4 +1,30 @@
-(* This file was warnings.ml from Objective Caml *)
+(**********************************************************************)
+(*                                                                    *)
+(*                           ReactiveML                               *)
+(*                    http://reactiveML.org                           *)
+(*                    http://rml.inria.fr                             *)
+(*                                                                    *)
+(*                          Louis Mandel                              *)
+(*                                                                    *)
+(*  Copyright 2002, 2007 Louis Mandel.  All rights reserved.          *)
+(*  This file is distributed under the terms of the Q Public License  *)
+(*  version 1.0.                                                      *)
+(*                                                                    *)
+(*  ReactiveML has been done in the following labs:                   *)
+(*  - theme SPI, Laboratoire d'Informatique de Paris 6 (2002-2005)    *)
+(*  - Verimag, CNRS Grenoble (2005-2006)                              *)
+(*  - projet Moscova, INRIA Rocquencourt (2006-2007)                  *)
+(*                                                                    *)
+(**********************************************************************)
+
+(* file: warnings.ml *)
+
+(* Warning: *)
+(* This file is based on the original version of warnings.ml *)
+(* from the Objective Caml 3.07 distribution, INRIA *)
+
+(* first modification: 2004-05-06  *)
+(* modified by: Louis Mandel *)
 
 (***********************************************************************)
 (*                                                                     *)
@@ -12,7 +38,7 @@
 (*                                                                     *)
 (***********************************************************************)
 
-(* $Id: warnings.ml,v 1.17 2003/05/02 12:52:11 weis Exp $ *)
+(* $Id$ *)
 
 (* Please keep them in alphabetical order *)
 
