--- asttypes.mli.org	2007-01-28 17:19:30.971017982 +0100
+++ ../global/asttypes.mli	2007-01-28 17:19:30.597067499 +0100
@@ -1,4 +1,30 @@
-(* This file was asttypes.ml from Objective Caml *)
+(**********************************************************************)
+(*                                                                    *)
+(*                           ReactiveML                               *)
+(*                    http://reactiveML.org                           *)
+(*                    http://rml.inria.fr                             *)
+(*                                                                    *)
+(*                          Louis Mandel                              *)
+(*                                                                    *)
+(*  Copyright 2002, 2007 Louis Mandel.  All rights reserved.          *)
+(*  This file is distributed under the terms of the Q Public License  *)
+(*  version 1.0.                                                      *)
+(*                                                                    *)
+(*  ReactiveML has been done in the following labs:                   *)
+(*  - theme SPI, Laboratoire d'Informatique de Paris 6 (2002-2005)    *)
+(*  - Verimag, CNRS Grenoble (2005-2006)                              *)
+(*  - projet Moscova, INRIA Rocquencourt (2006-2007)                  *)
+(*                                                                    *)
+(**********************************************************************)
+
+(* file: asttypes.mli *)
+
+(* Warning: *)
+(* This file is based on the original version of asttypes.mli *)
+(* from the Objective Caml 3.07 distribution, INRIA *)
+
+(* first modification: 2004-04-23 *)
+(* modified by: Louis Mandel      *)
 
 (***********************************************************************)
 (*                                                                     *)
@@ -12,27 +38,26 @@
 (*                                                                     *)
 (***********************************************************************)
 
-(* $Id: asttypes.mli,v 1.9 2003/04/25 12:27:30 xleroy Exp $ *)
-
-(* Auxiliary a.s.t. types used by parsetree and typedtree. *)
+(* $Id$ *)
 
-type constant =
-    Const_int of int
-  | Const_char of char
-  | Const_string of string
-  | Const_float of string
-  | Const_int32 of int32
-  | Const_int64 of int64
-  | Const_nativeint of nativeint
+(* Types used in the a.s.t. *)
 
-type rec_flag = Nonrecursive | Recursive | Default
+type rec_flag = Nonrecursive | Recursive
 
 type direction_flag = Upto | Downto
 
-type private_flag = Private | Public
+type immediate_flag = Immediate | Nonimmediate
 
-type mutable_flag = Immutable | Mutable
+type mutable_flag = Mutable | Immutable
 
-type virtual_flag = Virtual | Concrete
+type await_kind = All | One
 
-type label = string
+type pre_kind = Status | Value
+
+type immediate =
+  | Const_unit
+  | Const_bool of bool
+  | Const_int of int
+  | Const_float of float
+  | Const_char of char
+  | Const_string of string
