--- linenum.mll.org	2007-01-04 23:48:02.000000000 +0100
+++ ../parsing/linenum.mll	2007-01-04 23:48:02.000000000 +0100
@@ -1,4 +1,30 @@
-(* This file was linenum.mll from Objective Caml *)
+(**********************************************************************)
+(*                                                                    *)
+(*                           ReactiveML                               *)
+(*                    http://reactiveML.org                           *)
+(*                    http://rml.inria.fr                             *)
+(*                                                                    *)
+(*                          Louis Mandel                              *)
+(*                                                                    *)
+(*  Copyright 2002, 2007 Louis Mandel.  All rights reserved.          *)
+(*  This file is distributed under the terms of the Q Public License  *)
+(*  version 1.0.                                                      *)
+(*                                                                    *)
+(*  ReactiveML has been done in the following labs:                   *)
+(*  - theme SPI, Laboratoire d'Informatique de Paris 6 (2002-2005)    *)
+(*  - Verimag, CNRS Grenoble (2005-2006)                              *)
+(*  - projet Moscova, INRIA Rocquencourt (2006-2007)                  *)
+(*                                                                    *)
+(**********************************************************************)
+
+(* file: linenum.mll *)
+
+(* Warning: *)
+(* This file is based on the original version of linenum.mll *)
+(* from the Objective Caml 3.07 distribution, INRIA          *)
+
+(* first modification: 2004-05-06 *)
+(* modified by: Louis Mandel      *)
 
 (***********************************************************************)
 (*                                                                     *)
@@ -12,7 +38,7 @@
 (*                                                                     *)
 (***********************************************************************)
 
-(* $Id: linenum.mll,v 1.7 1999/11/17 18:58:17 xleroy Exp $ *)
+(* $Id$ *)
 
 (* An auxiliary lexer for determining the line number corresponding to
    a file position, honoring the directives # linenum "filename" *)
@@ -37,7 +63,7 @@
     if !f1 < String.length s then
       filename := String.sub s (!f1 + 1) (!f2 - !f1 - 1)
   with Failure _ | Invalid_argument _ ->
-    Misc.fatal_error "Linenum.parse_sharp_line"
+    failwith "Linenum.parse_sharp_line"
 }
 
 rule skip_line = parse
